File: utils/Operators/UPDATE_SOUND_SPEED.jl
function UPDATE_SOUND_SPEED(s::Dict{String, Any}, chemistry::Dict{String, Any})
# UPDATE_SOUND_SPEED - Compute the local speed of sound from the current flow state.
#
# Evaluates the speed of sound either from chemistry lookup tables
# or from the ideal-gas relation a = sqrt(gamma* * p / rho).
#
# Part of: Hypersonics Stability Julia Solver - Operators Module

    ## Extract dict refs
    var  = s["var"]::Dict{String, Any}
    chem = s["chemistry"]::Dict{String, Any}
    fs   = s["freestream"]::Dict{String, Any}

    ## Compute internal energy
    rho       = var["rho"]::Matrix{Float64}
    rho_E     = var["rho_E"]::Matrix{Float64}
    rho_u_arr = var["rho_u"]::Matrix{Float64}
    rho_v_arr = var["rho_v"]::Matrix{Float64}
    e = @. (rho_E - (rho_u_arr^2 + rho_v_arr^2) / rho / 2) / rho

    ## Update sound speed
    rho_factor    = fs["rho_factor"]::Float64
    energy_factor = fs["energy_factor"]::Float64
    vel_factor    = fs["velocity_factor"]::Float64

    if chem["is_chemistry_enabled"]::Bool
        eval_a = chemistry["eval_a"]
        rho_dim = @. rho_factor * rho
        e_dim   = @. energy_factor * e
        var["a"] = @. eval_a(rho_dim, e_dim) / vel_factor
    else
        p          = var["p"]::Matrix{Float64}
        gamma_star = var["gamma_star"]::Matrix{Float64}
        var["a"] = @. sqrt(gamma_star * p / rho) / vel_factor
    end

    return s
end

File: utils/Operators/UPDATE_SOUND_SPEED.m
function s = UPDATE_SOUND_SPEED(s, chemistry)
% UPDATE_SOUND_SPEED  Compute the local speed of sound from the current flow state.
%
%   s = UPDATE_SOUND_SPEED(s, chemistry)
%
%   Evaluates the speed of sound either from chemistry lookup tables
%   (when chemistry is active) or from the ideal-gas relation
%   a = sqrt(gamma* * p / rho) for a calorically perfect gas.
%
%   Inputs:
%       s  - Solution struct containing:
%                   .var.rho, .var.rho_u, .var.rho_v, .var.rho_E - Conserved variables
%                   .chemistry.is_chemistry_enabled - Logical, true if chemistry active
%                   .var.gamma_star      - Ratio of specific heats field
%                   .var.p               - Pressure field
%                   .freestream.*        - Scaling factors (rho_factor,
%                                          energy_factor, velocity_factor)
%       chemistry - Chemistry model struct with:
%                   .eval_a - Speed of sound evaluation function(rho, e)
%
%   Outputs:
%       s  - Updated s struct with:
%                   .var.a - Speed of sound field (non-dimensional)
%
%   Notes:
%       - Internal energy is computed as e = (rho_E - KE) / rho.
%       - For chemistry cases, the dimensional result from eval_a is
%         non-dimensionalized by velocity_factor.
%
% Part of: Hypersonics Stability MATLAB Solver - Operators Module

    %% Compute internal energy
    rho = s.var.rho;
    e = (s.var.rho_E - (s.var.rho_u.^2 + s.var.rho_v.^2) ./ rho / 2) ./ rho;

    %% Update sound speed
    if s.chemistry.is_chemistry_enabled
        s.var.a = chemistry.eval_a(s.freestream.rho_factor * rho, s.freestream.energy_factor * e) / s.freestream.velocity_factor;
    else
        s.var.a = sqrt(s.var.gamma_star .* s.var.p ./ s.var.rho) / s.freestream.velocity_factor;
    end
end
